database/misc/couchdb_user

```diff
@@ -0,0 +1,315 @@
+      - derived_key: hash of the password.
+      - salt: salt used when hashing the password.
+      - iterations: iterations of PBKDF2 run. CouchDB uses 10 by default.
+
+      When password field is posted, CouchDB automatically updates derived_key.
+      The plaintext is never stored.
+    """
+
+    # http://docs.couchdb.org/en/latest/intro/security.html#creating-new-user
+    url = couchdb_user_url(host, username)
+
+    if rev is not None:
+        url = '%s?rev=%s' % (url, rev)
+
+    if roles is None:
+        roles = []
+
+    if type(roles) not in (list, tuple):
+        raise TypeError("roles should be a list or tuple")
+
+    data = {
+        'name': username,
+        'roles': roles,
+        'type': 'user',
+    }
+
+    if password is not None and derived_key is not None:
+        raise ValueError("pass either password or derived_key")
+
+    if iterations is None:
+        iterations = 10
+
+    authdata = (derived_key, salt)
+    if any(authdata) and not all(authdata):
+        raise ValueError("one of derived_key or salt is missing")
+
+    if password is not None:
+        data['password'] = password
+    elif derived_key is not None:
+        data['derived_key'] = derived_key
+        data['salt'] = salt
+        data['iterations'] = iterations
+
+    r = requests.put(url, auth=credentials, data=json.dumps(data))
+    r.raise_for_status()
+    return True
+
+
+def couchdb_user_valid_password(plaintext, salt, hash_, rounds=10):
+    """
+    Check if a given plaintext and salt corresponds to a given hash
+    """
+
+    if type(salt) == unicode:
+        salt = salt.encode('utf-8')
+    hashed = pbkdf2(plaintext, salt, rounds)
+    return hexlify(hashed) == hash_
+
+
+# ===========================================
+# Module execution.
+#
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            login_user=dict(default=None),
+            login_password=dict(default=None),
+            login_host=dict(default="localhost"),
+            login_port=dict(default="5984"),
+
+            name=dict(required=True, aliases=['username']),
+            password=dict(required=False),
+            roles=dict(required=False, type='list', default=[]),
+            state=dict(default="present", choices=["absent", "present"]),
+        )
+    )
+
+    username = module.params["name"]
+    password = module.params["password"]
+    roles = module.params["roles"]
+    state = module.params["state"]
+
+    if roles is None:
+        roles = []
+
+    # see https://github.com/ansible/ansible/issues/9254
+    if roles == ['']:
+        roles = list()
+
+    login_password = module.params['login_password']
+    login_user = module.params['login_user']
+
+    host = (
+        module.params['login_host'],
+        module.params['login_port']
+    )
+
+    credentials = (
+        login_user,
+        login_password
+    )
+
+    if any(credentials):
+        if not all(credentials):
```

Comment:
Its ok but I wanted to let you know there is a helper for arguments required together here https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L512 e.g. look for required_together= in other modules how to use it if you like